### ISSUE TITLE
requirments of pointers in stacked borrows

### ISSUE THREAD

**Issue by CeleritasCelery** [rust-lang/unsafe-code-guidelines#324]
I have been thinking about [this comment](https://github.com/rust-lang/unsafe-code-guidelines/issues/256#issuecomment-1076555084) and it's implication for pointers. Not sure if this discussed else where, but I couldn't find the issue. Anyways here is the reference code:

```rust
unsafe fn uwu(ptr: *mut i32) {
  let ref1 = &mut *ptr;
  let ptr1 = ref1 as *mut _;
  let ref2 = &mut *ptr;
  let ptr2 = ref2 as *mut _;

  // There is UB here
  *ptr1 = 0; *ptr2 = 0; *ptr1 = 0;
}
```

I don't understand why this is UB. My best understanding is that when we write through `ptr1`, we assert uniqueness of that pointer, then writing though `ptr2` (which came from a different reference) violates that? Is the same reason this code is not allowed? 
```rust
let ref1: &mut i32 = ...
let ref2: &mut i32 = ref1;
*ref1 = 0;
*ref2 = 0;
```

I am trying to understand this new semantic property of pointers. As I understand it, pointers have 4 properties.

- validity (point to initialized memory)
- alignment
- address
- provenance

if I have a function like this:
```rust
fn foo(a: *mut T, b: *mut T) {}
```

If `a` and `b` have the same address I know that they are essentially equivalent in that:
1. If one is valid they are both valid
2. If one is aligned they are both aligned
3. They both have provenance of at least `T`

For all intents and purposes I can treat these pointers the same. However with this new semantic property (I will call it "stacked uniqueness"). That no longer holds true. If even if they have the same address, I cannot treat them interchangeably. There is this new property that I can't reason about unless I know how the pointers were obtained. And the order I use the pointers internally in my function now needs to be exposed to my caller so as to prevent UB. And an internal refactor could lead to UB for the caller. Unlike with references, Rust provides no guarantees or help with correct "stacked uniqueness" of pointers. But it is a requirement none the less. I feel like this could damage local reasoning of code.

Of course, I don't really understand this new property of pointers other then that it exists and it is derived from where pointers are created. So it may not be as bad as I think. Are the details of this "more permissive" stacked borrows documented anywhere? Currently with Miri and `-Zmiri-tag-raw-pointers` stacked borrows will essentially treat pointers as references (in that they need to have unique tags in the borrow stack to mutate things). I don't know how much the "permissive model" relaxes that.

To summarize, the questions being asked here is what are the requirements of "stacked uniqueness" of pointers in a "permissive model" of stacked borrows? How do they impact local reasoning of code? How much are these needed for optimizations of references?

**Comment by RalfJung** [comment]
> I don't understand why this is UB.

Currently (under Stacked Borrows), this is UB because when you do `let ref2 = &mut *ptr;` you are saying "`ref2` is a unique pointer, and no other pointer currently exists that points to the same thing". This invalidates `ref1` (which the borrow checker knows, so it will prevent you from using it) and `ptr1`.

> Are the details of this "more permissive" stacked borrows documented anywhere?

No, that model is a vague set of ideas in my head that might or might not actually work. It's not a model that exist yet, it's a hypothetical future model.

In that model I think a lot of the "ancestry tree" of pointers and references. For the example, that tree is something like
```
    ptr
  /    \
ref1  ref2
  |    |
ptr1  ptr2
```
Here, `ref1` and `ref2` are "unique" nodes since they don't allow aliasing. In my vague idea, whenever a unique pointer *or any of its descendants* is used, then all "sibling" pointers get invalidated. So, the moment `ptr1` is used, all siblings of `ref1` get invalidated, which means `ref2` and all its descendents including `ptr2`.

My intuition is fairly strong that if we don't do this, we will have much fewer actionable aliasing guarantees. I am also pretty sure that we have to do this if we want to be compatible with LLVM `noalias`, which is intended to model C `restrict` and that one certainly always talks about a pointer "and all pointers derived from it" (i.e., its descendents).

But as I said this is all a vague soup in my head, nothing is set in stone here and I have not thought about this enough to actually give concrete examples for what we would lose if we did not do it this way.
